course-material/interactive-queries-and-parameterization/make-your-queries-interactive.md

```diff
@@ -1,14 +1,471 @@
+Consider the `pdays` variable from the bank marketing dataset. The value in this column is -1 when the client was not contacted since the previous campaign and an integer > 0 otherwise. The marketing manager wants you to create a macro named `dummify` that transforms this numeric variable into a binary categorical variable, named `pdays_dummy`, taking on either "no" if `pdays` = -1 or "yes" otherwise. You are also expected to create both a `RadioButtons` selection widget for the transformed categorical variable and a `SelectMultiple` selection widget for the `poutcome` variable to help the manager filter for campaign performance on the fly. Finally, output the rendered query after displaying the tabular results.
+
+<b>Hint</b> Creating selection widgets within a macro is syntactically different from creating numeric widgets. To do so, we need to use a `WITH` clause to create a temporary relation, only available in the query. Then, use the temporary relation to call the widget variables in a new SELECT ... FROM ... WHERE ... query. Nested subqueries would work but also make the query hard to read and understand. See the `WITH` clause's documentation [here](https://duckdb.org/docs/sql/query_syntax/with.html) for more details.
+
+<!-- #region -->
+<details>
+<summary>Answers</summary>
+
+We start off by initializing our selection widget variables:
+
+```{code-cell} ipython3
+contact_selection = widgets.RadioButtons(
+    options=["yes", "no"], description="Previously Contacted?", disabled=False
+)
+```
+
+```{code-cell} ipython3
+outcome_selection = widgets.SelectMultiple(
+    options=["failure", "other", "success", "unknown"],
+    value=["success", "failure"],
+    description="Campaign Outcome:",
+    disabled=False,
+)
+```
+
+Now comes the hard part! Because we are creating a macro, we need to use the `--save` argument for it as well as the `--interact` argument for our widget variables. We then proceed to create the discretization macro function. Here, because we want to output "yes" or "no", using `::numeric(16, {{ precision }}` after the `CASE WHEN` statement will be incorrect; hence, we use `::varchar`. 
+
+The `WITH` clause then helps us create our temporary relation, named `transformed_bank`, with all the columns, including the transformed `pdays_dummy`, we want to output for our query. Lastly, a new query uses the columns from the temporary relation, helping us effectively use the widgets with the help of its `WHERE` clause: 
+
+```{code-cell} ipython3
+%%sql --save dummify --interact contact_selection --interact outcome_selection
+{% macro days_to_dummy(column_name) %}
+    (case when {{ column_name }} = -1 then 'no' else 'yes' end)::varchar
+{% endmacro %}
+
+WITH transformed_bank AS (
+  SELECT
+    job,
+    marital,
+    poutcome,
+    {{ days_to_dummy('pdays') }} as pdays_dummy
+  FROM bank
+)
+SELECT *
+FROM transformed_bank
+WHERE poutcome IN {{outcome_selection}} AND 
+pdays_dummy == '{{contact_selection}}';
+```
```

Comment:
Yep I share that concern as well. Moreover, we did not cover the `WITH` clause in the following module, so I referenced DuckDB's docs for more info on it. I tried adding as deep an explanation for using `WITH` while keeping it brief and if we keep it, it would do more good than bad I feel for new students

Reply:
Another option is to use the more simple solution

```
SELECT
  job, marital, 
  {{ days_to_dummy('pdays') }} as pdays_dummy
FROM bank
WHERE {{ days_to_dummy('pdays') }} == '{{contact_selection}}';

```

Reply:
Full solution

```
%%sql --save dummify --interact contact_selection --interact outcome_selection
{% macro days_to_dummy(column_name) %}
    (case when {{ column_name }} = -1 then 'no' else 'yes' end)::varchar
{% endmacro %}
SELECT
  job, marital, 
  {{ days_to_dummy('pdays') }} as pdays_dummy
FROM bank
WHERE {{ days_to_dummy('pdays') }} == '{{contact_selection}}';
```